File: rust/src/leet_code/988.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

use std::rc::Rc;
use std::cell::RefCell;

/// This module contains the implementation of the `Solution` struct.
///
/// The `Solution` struct provides two methods: `dfs` and `smallest_from_leaf`.
/// The `dfs` method is a helper function used to perform a depth-first search on a binary tree.
/// The `smallest_from_leaf` method returns the lexicographically smallest string that starts at a leaf of this tree and ends at the root.
///
/// # Structs
///
/// * `Solution` - The main struct in this module.
///
/// # Methods
///
/// * `dfs` - Performs a depth-first search on a binary tree.
/// * `smallest_from_leaf` - Returns the lexicographically smallest string that starts at a leaf of this tree and ends at the root.
impl Solution {
    /// Performs a depth-first search on a binary tree.
    ///
    /// # Parameters
    ///
    /// * `node` - A reference to an optional `Rc<RefCell<TreeNode>>`.
    /// * `res` - A mutable reference to a `String`.
    /// * `path` - A mutable `String`.
    ///
    /// # Returns
    ///
    /// This method does not return a value.
    pub fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, res: &mut String, mut path: String) {
        // If the node is Some, borrow it
        if let Some(node) = node {
            let node = node.borrow();
            // Push the node value to the path
            path.push((node.val as u8 + b'a') as char);
            // If the node is a leaf node
            if node.left.is_none() && node.right.is_none() {
                // Reverse the path and collect it into a String
                let path = path.chars().rev().collect::<String>();
                // If res is empty or path is lexicographically smaller than res, assign path to res
                if res.is_empty() || path < *res {
                    *res = path;
                }
            }
            // Recursively call dfs on the left and right child nodes
            Self::dfs(&node.left, res, path.clone());
            Self::dfs(&node.right, res, path);
        }
    }

    /// Returns the lexicographically smallest string that starts at a leaf of this tree and ends at the root.
    ///
    /// # Parameters
    ///
    /// * `root` - An optional `Rc<RefCell<TreeNode>>`.
    ///
    /// # Returns
    ///
    /// * `String` - The lexicographically smallest string that starts at a leaf of this tree and ends at the root.
    pub fn smallest_from_leaf(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        // Initialize res as an empty String
        let mut res = String::new();
        // Call dfs on the root node
        Self::dfs(&root, &mut res, String::new());
        // Return res
        res
    }
}

File: cpp/leet_code/988.cpp
// Optimizing the code for speed and unrolling loops for efficiency.
#pragma GCC optimize("O3,unroll-loops")
#include <iostream>
#include <string>

// Solution class
class Solution {
public:
    // Member variable to store the answer
    std::string ans;

    /**
     * Function to find the smallest string from leaf to root in a binary tree.
     * @param node: Pointer to the current TreeNode.
     * @param temp: Reference to the string representing the current path.
     */
    void FindSmallest(const TreeNode *node, std::string &temp) {
        // Append the current character represented by the node's value to the path.
        temp += ('a' + node->val);

        // If the current node is a leaf node
        if (!node->left && !node->right) {
            // Reverse the current path to get the path from leaf to root.
            const std::string final_string(temp.rbegin(), temp.rend());

            // If ans is empty or final_string is smaller than ans, update ans.
            if (ans.empty() || final_string < ans) {
                ans = final_string;
            }
        }

        // If the current node has a left child, continue the traversal on the left subtree.
        if (node->left) {
            FindSmallest(node->left, temp);
            temp.pop_back(); // Backtrack
        }

        // If the current node has a right child, continue the traversal on the right subtree.
        if (node->right) {
            FindSmallest(node->right, temp);
            temp.pop_back(); // Backtrack
        }
    }

    /**
     * Function to find the lexicographically smallest string from leaf to root in a binary tree.
     * @param root: Pointer to the root of the binary tree.
     * @return The lexicographically smallest string from leaf to root.
     */
    std::string smallestFromLeaf(const TreeNode *root) {
        // Improve the performance of cin
        std::ios_base::sync_with_stdio(false);
        std::cin.tie(nullptr);

        // String to store the current path
        std::string temp;

        // Start the traversal from the root
        FindSmallest(root, temp);

        // Return the smallest string from leaf to root
        return ans;
    }
};
